sysbuild/CMakeLists.txt

```diff
@@ -0,0 +1,76 @@
+# Copyright (c) 2021 Nordic Semiconductor
+#
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20)
+
+if(NOT DEFINED APP_DIR)
+  message(FATAL_ERROR "No main application specified")
+endif()
+
+# Add sysbuild/cmake/modules to CMAKE_MODULE_PATH which allows us to integrate
+# sysbuild CMake modules with general Zephyr CMake modules.
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake/modules)
+# List of Zephyr and sysbuild CMake modules we need for sysbuild.
+# Note: sysbuild_kconfig will internally load kconfig CMake module.
+set(zephyr_modules extensions sysbuild_extensions python west root zephyr_module boards shields sysbuild_kconfig)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE} COMPONENTS ${zephyr_modules})
+
+project(sysbuild LANGUAGES)
+
+# Global list of images enabled in this multi image build system.
+set(IMAGES)
+
+get_filename_component(APP_DIR  ${APP_DIR} ABSOLUTE)
+get_filename_component(app_name ${APP_DIR} NAME)
+
+# Propagate bootloader and signing settings from this system to the MCUboot and
+# application image build systems.
+if(SB_CONFIG_BOOTLOADER_MCUBOOT)
+  set(${app_name}_CONFIG_BOOTLOADER_MCUBOOT y CACHE STRING
+      "MCUBOOT is enabled as bootloader" FORCE
+  )
+  set(${app_name}_CONFIG_MCUBOOT_SIGNATURE_KEY_FILE
+      \"${SB_CONFIG_BOOT_SIGNATURE_KEY_FILE}\" y CACHE STRING
+      "Signature key file for signing" FORCE
+  )
+
+  # Set corresponding values in mcuboot
+  set(mcuboot_CONFIG_BOOT_SIGNATURE_TYPE_${SB_CONFIG_SIGNATURE_TYPE} y CACHE STRING
+      "MCUBOOT signature type" FORCE
+  )
+  set(mcuboot_CONFIG_BOOT_SIGNATURE_KEY_FILE
+      \"${SB_CONFIG_BOOT_SIGNATURE_KEY_FILE}\" y CACHE STRING
+      "Signature key file for signing" FORCE
+  )
+else()
+  set(${app_name}_CONFIG_BOOTLOADER_MCUBOOT n CACHE STRING
+      "MCUBOOT is disabled as bootloader" FORCE
+  )
+endif()
+
+# This adds the primary application to the build.
+ExternalZephyrProject_Add(
+  APPLICATION ${app_name}
+  SOURCE_DIR ${APP_DIR}
+  MAIN_APP
+)
+list(APPEND IMAGES "${app_name}")
+set(DEFAULT_IMAGE "${app_name}")
+
+# Include MCUboot if enabled.
+if(SB_CONFIG_BOOTLOADER_MCUBOOT)
+  ExternalZephyrProject_Add(
+    APPLICATION mcuboot
+    SOURCE_DIR ${ZEPHYR_MCUBOOT_MODULE_DIR}/boot/zephyr/
+  )
+  # MCUBoot should be first in list, as it is a bootloader.
+  list(PREPEND IMAGES "mcuboot")
+endif()
+
+# This allows for board and app specific images to be included.
+include(${BOARD_DIR}/sysbuild.cmake OPTIONAL)
```

Comment:
would SOC specific images be possible in this same block using `SOC_DIR`? One use case for this would be booting the secondary core of a multicore SOC using a shim application running on the primary.

Reply:
Yep, that's a good idea.

And with the merge of #41301 this should now be possible.
(was not possible in the initial prototype because of the old boilerplate.cmake design)

Will take a look at this.

Reply:
Update.

Although it's a good idea then we need to wait a little longer with SoC support.

The reason is that SoC relies on `SOC_NAME`, `SOC_SERIES`, and `SOC_FAMILY`:
https://github.com/zephyrproject-rtos/zephyr/blob/d74cb2a3df2376dd65936d30cf4dc56092044761/cmake/modules/soc.cmake#L36-L39

So although the improvements from #41301 takes us part of the way then we still rely on Kconfig parsing from the regular Zephyr Kconfig tree.

And we don't want all the Kconfig tree of Zephyr parsed in the sysbuild itself (risk of creating a nightmare of `depends on` / `select` we would need to handle).

Will keep in mind, and if I come up with a clever / minimal solution, then I will update this PR.
The request is being tracked here: #43488